More backend and frontend integrations

nmdanny opened the pull request:
Still WIP, but here's what I'm trying to do

- [x] Connect user settings to backend

- [x] Add a courses by track API (e.g `/tracks/1337/courses` would return all courses which are included in any CourseGroup that involve track with pk 1337), and integrate it into FE.

- [ ] Optimize scrape script (retrieval of courses)

- Other minor FE/BE UI fixes

- [ ] Might try replacing the `SearchBar` component with something from a proper library, or leave it to a different PR?

nmdanny committed:
temporary refactor that involves search-bar

will most likely replace it alltogether

nmdanny committed:
WIP FE settings

nmdanny committed:
minor BE auth0 changes

Minor changes to be inline with Auth0 tutorial
for Django

nmdanny committed:
Settings page can now save the user track

Note that this deletes the user's courses for now, but they aren't used anyway.

nmdanny committed:
'App' is solely responsible for getting student & claims

nmdanny committed:
Do not include course-groups in TrackViewSet list

Including course-groups massively increases the size of
the returned tracks

nmdanny committed:
add courses by track route, use it in FE(WIP)

nmdanny committed:
fix warnings and test

nmdanny committed:
add course type to huji import + add more coursetype colors

ScDor commented on Closure_Front_End/src/components/Navigation.vue:
```diff
@@ -1,16 +1,20 @@
-  <year-selection v-model="selectedYear" />
+  <year-selection label="גרסת שנתון" v-model="selectedYear" />
   <search-bar
     :placeholder="'חפש מסלול'"
```
```suggestion
    :placeholder="'חיפוש מסלול'"
```

gender-neutral

ScDor commented on Closure_Front_End/src/components/Navigation.vue:
```diff
@@ -1,16 +1,20 @@
   ></search-bar>
 
   <search-bar
     :placeholder="'חפש קורס'"
```
```suggestion
    :placeholder="'חיפוש קורס'"
```

ScDor commented on Closure_Front_End/src/components/User.vue:
```diff
@@ -1,32 +1,49 @@
+
         <li>
+
+          <label class="menu-label">מחזור(שנתון)</label>
```
is this the year when the student started? if so, i'd call it שנת תחילת לימודים (אוקטובר של..)

ScDor commented on Closure_Front_End/src/components/User.vue:
```diff
@@ -1,32 +1,49 @@
         <li>
           <div class="control">
-            <button class="button menu-label is-dark">שמור</button>
+            <button :disabled="!newTrack" class="button menu-label is-dark" :class="{'is-loading': saving}" @click="$emit('onSave', $event, newTrack)">שמור</button>
```
```suggestion
            <button :disabled="!newTrack" class="button menu-label is-dark" :class="{'is-loading': saving}" @click="$emit('onSave', $event, newTrack)">שמירה</button>
```

ScDor commented on Closure_Front_End/src/huji-import/CandidateCourses.vue:
```diff
@@ -17,12 +17,20 @@
     מהקורסים, יש לבחור ידנית את הסמסטר
   </div>
+  <div class="notification is-warning" v-if="!isTrackDefined">
+    לא הוגדר מסלול עבור המשתמש, אז סוגי הקורסים אינם מוגדרים.
```
```suggestion
    לא הגדרת מסלול לימודים, אז אנחנו לא יכולים להסיק את סוגי הקורסים (חובה, חובת בחירה..) - מומלץ לבחור מסלול בצד ימין כדי לתכנן את הלימודים.
```

ScDor commented on Closure_Front_End/src/huji-import/CandidateCourses.vue:
```diff
@@ -17,12 +17,20 @@
+    לא הוגדר מסלול עבור המשתמש, אז סוגי הקורסים אינם מוגדרים.
+  </div>
+  <div class="notification" v-if="isTrackDefined">
+    סוגי הקורסים הם ביחס למסלול המוגדר אצל המשתמש, 
```
```suggestion
    סוגי הקורסים הם ביחס למסלול הלימודים שבחרת
```

ScDor commented on Closure_Front_End/src/huji-import/CandidateCourses.vue:
```diff
@@ -83,6 +92,15 @@
+const MODEL_COURSE_TYPE_TO_STRING = new Map([
+  ["MUST", "חובה"],
+  ["CHOICE", "בחירה"],
+  ["CHOOSE_FROM_LIST", "בחירה מרשימה"],
```
```suggestion
  ["CHOOSE_FROM_LIST", "חובת בחירה"],
```

ScDor commented on Closure_Front_End/src/huji-import/CandidateCourses.vue:
```diff
@@ -83,6 +92,15 @@
+  ["CHOICE", "בחירה"],
+  ["CHOOSE_FROM_LIST", "בחירה מרשימה"],
+  ["CORNER_STONE", "אבן פינה"],
+  ["SUPPLEMENTARY", "משלים"]
```
```suggestion
  ["SUPPLEMENTARY", "לימודים משלימים"]
```

ScDor commented on Closure_Front_End/src/components/CourseBox.vue:
```diff
@@ -28,8 +28,10 @@ export default {
+      if (this.course.type === "CHOOSE_FROM_LIST") return "choose_from_list";
+      if (this.course.type === "CORNER_STONE") return "corner_stone";
+      if (this.course.type === "SUPPLEMENTARY") return "supplementary";
       return "choice";
```
Supplementary should be the default, and choice should only be used if the type is choice

nmdanny committed:
Adjust colors

nmdanny committed:
Change phrasing in some places

nmdanny committed:
get rid of unused Home code

retreival of student courses is currently done from local storage,
and will be done in a more organized future for DB support in the future

nmdanny committed:
Use better select component, refactor course navigation

There are some style issues(RTL) with the new component, and spacing
issues in the user settings page, hopefully these can be fixed

nmdanny committed:
WIP add course detail, refactor Home

- Create CourseDetail component, which will be mostly used for when adding a course
- Refactor ProgressBox outside of Home
- 'Home' is now responsible for the currently selected course, you can select
  a course either by clicking on it(From one of the 'Year' components) or
  via the select at 'Navigation' component)

nmdanny committed:
use kebab case in FE components

ScDor commented:
Can we _show_ the options for data-years as `2021-2022`, so they're clearer to the user? (while, behind the scenese, using the value as `2022`)

nmdanny committed:
refactoring of Take and Student, add CoursePlan

- A student is no longer associated to courses directly.
  Instead, he can define one or more course-plans, which can
  also be shared(made public)

- Implement object level auth for course plans - only the owner
  can access it, unless the plan was made public, then it can be read
  by all users(as well as unauthenticated ones), but it cannot be
  modified by them.

- Add tests for the new course plan serializer, as well as CoursePlan views and
  permissions

- Student.year_in_studies is now calculated based on his chosen track
  year*

- Student.remaining and Student.trickle down are temporarily disabled*

* Perhaps these calculations should be moved to be someone else's
responsibility, either into CoursePlan or a new API?

nmdanny commented:
I've been working on allowing multiple course-plans per user(and sharing them), as per #115 - currently backend only
(It might warrant a different PR, for now I've been on this PR draft to allow easier discussions)

If we allow for multiple plans, then the same user might try out plans for different tracks (for example, if he's contemplating which track to pick, or wants to help a friend in another faculty organize his plan) - each one of them would yield different analytics and point calculations. Therefore, properties like `Student.remaining/trickle_down` and `Student.track` should be moved to `CoursePlan`, in my opinion.

nmdanny committed:
Forgot to remove TakeSerializer (renamed to CoursePlanSerializer)

nmdanny committed:
move 'track' and 'remaining' from Student into CoursePlan

nmdanny committed:
Add "IsAdminOrAuthenticatedReadOnly" permission

This ensures ordinary users cannot modify tracks/courses,
but can still fetch them

nmdanny committed:
Show course type for courses by tracks endpoint only

type is no longer dependant on the authenticated student

nmdanny committed:
Take type depends on course_plan track, not student

nmdanny committed:
standardize take of a course in FE

- semester is now stored as a string(same enum as in BE), and not ints.
  This makes some utility functions redundant and simplifies processing.

- The 'semester' and 'year' fields of Course, have been moved to
   the 'take' object field of a Course. Now, the Course 'data_year'
  and 'semester' fields purely reflect the course's state in the DB.

  This distinction is important for courses with any "any" field,
  whose assigned take in pratice(sem1/sem2) may be different
  from the one stated in the DB, as well as annual semesters which
  are currently classified as year 1.

nmdanny committed:
Use 'id' instead of 'pk' in serializers

For consistency with model definitions

nmdanny committed:
Add FE support for saving a course plan

nmdanny committed:
Take/Courseplan retrieve endpoints include detail view for GET

The detail view contains entire course objects within the takes, and not just
their IDs - useful for loading a course plan

nmdanny committed:
fix save state footer

nmdanny committed:
Add 'swrv' data fetching library to FE

nmdanny committed:
Implement loading of course plans via FE

nmdanny committed:
Move save/load components to their own folder

nmdanny committed:
Refactor 'Home' to use composition API

nmdanny committed:
BE fix + show course type within courseplan

- DetailCoursePlanSerializer accidentally inherited from CourseSerializer rather
  than CoursePlanSerializer

- 'CourseOfTrackSerializer' now has the track passed by
  'DetailCoursePlanSerializer' (if a track is defined for the
   course plan)

nmdanny committed:
BE remove unnecessary line

nmdanny committed:
FE store fix typo

nmdanny committed:
FE - WIP add course plan settings component

nmdanny committed:
FE WIP adjustments to save modal

- add checkbox for 'public' option

- dont close modal immediately, maybe I'll add a visual indicator
  for a  successful save, or perhaps a link to the newly created plan
  for sharing/bookmarking?

nmdanny committed:
WIP load plan from link

nmdanny committed:
FE minor fixes regarding track shown after saving

nmdanny committed:
Use scoped CSS for a few components

The lack of scoping caused conflicts when importing these
component in certain order

nmdanny committed:
FE WIP more changes to loading

- Loading a plan through the UI uses the same mechanism
  as loading from a URL

- Now presents a warning when tried to load while there's an
  unsaved plan

- Path should reflect the current plan, even after saving or
  deleting the curret plan.

nmdanny committed:
Add copyable link to save as modal

nmdanny committed:
Remove unused stuff in Settings page (now useless)

nmdanny committed:
Refactor select components

- Use a better library for the select component, which
  seems to support RTL, is less buggy and more customizable

- Factor track selection and course selection to their own
  components(using said library)

- Year selection also uses said library, instead of a plain HTML5
   select.

- course selection is still WIP

nmdanny committed:
Refactor course color to component, use in course selection

nmdanny committed:
Some fixes regarding course plan navigation

nmdanny committed:
BE: undo useless max_limit check in pagination

The check was causing bad requests to cause an
internal server error(due to wrong exception being
thrown) rather than a HTTP 400.

Regardless, LimitOffsetPagination already enforces
the max_limit(by simply truncating the results and
adding a pointer to the next request).

nmdanny committed:
BE: add inclusion filters for Course viewsets

This will allow retrieving several courses with
a single request

nmdanny committed:
FE: add a cached course service

So far this service includes 2 operations

- Find all course instances of given course_id (that is, in which years
  is a given course offered)?

- Associate between a track and a list of courses, to their types.

These services  add an IndexedDB cache
(like local-storage, but more efficient and bigger) via the 'Dexie'
library, which will eliminate a bunch of network calls.

nmdanny committed:
add hook for syncing course types from current track

nmdanny committed:
refactor course detail to show all possible years
